File: src/haskell/Loc.hs
module Loc (
    Loc(..),
    newLoc,
    distance,
    normal
) where

data Loc = Loc Int Int Int deriving (Show, Eq)

newLoc :: Int -> Int -> Int -> Loc
newLoc x y z = normal (Loc x y z) (Loc 0 0 0)

distance :: Loc -> Loc -> Int
distance l1 l2 =
    let (Loc x y z) = normal l1 l2 in
    (abs x) + (abs y) + (abs z)

normal :: Loc -> Loc -> Loc
normal (Loc x y z) (Loc ox oy oz) =
    let (rx, ry, rz) = (x - ox, y - oy, z - oz)
        (ax, ay, az) = (abs rx, abs ry, abs rz)
        delta = getDelta rx ry rz ax ay az in
    (Loc (x - delta) (y + delta) (z - delta))


getDelta :: Int -> Int -> Int -> Int -> Int -> Int -> Int
getDelta rx ry rz ax ay az =
    -- I && II
    -- +x +y +z
    if rx >= 0 && ry >= 0 && rz >= 0 then
        nearZero rx rz
    -- II && III
    -- -x +y +z
    else if rx <= 0 && ry >= 0 && rz >= 0 then
        let amount = nearZero rx ry in
        if ax > ay then
            -amount
        else if ax == ay then
            -abs amount
        else
            amount
    -- III && IV
    -- -x -y +z
    else if rx <= 0 && ry <= 0 && rz >= 0 then
        let amount = nearZero ry rz in
        if ay < az then
            -amount
        else if ay == az then
            abs amount
        else
            amount
    -- IV && V
    -- -x -y -z
    else if rx <= 0 && ry <= 0 && rz <= 0 then
        nearZero rx rz
    -- V && VI
    -- +x -y -z
    else if rx >= 0 && ry <= 0 && rz <= 0 then
        let amount = nearZero rx ry in
        if ax > ay then
            -amount
        else if ax == ay then
            abs amount
        else
            amount
    -- VI && I
    -- +x +y -z
    else if rx >= 0 && ry >= 0 && rz <= 0 then
        let amount = nearZero ry rz in
        if ay < az then
            -amount
        else if ay == az then
            -abs amount
        else
            amount
    -- +x -y +z
    else if rx >= 0 && ry <= 0 && rz >= 0 then
        -- +x axis
        if ax > az && az == ay then
            abs $ nearZero ry rz
        -- I
        else if ax > az && az > ay then
            az
        -- +y axis
        else if az == ax && ax > ay then
            abs $ nearZero rx rz
        -- II
        else if az > ax && ax > ay then
            ax
        -- +z axis
        else if az > ay && ay == ax then
            abs $ nearZero rx ry
        -- III
        else if az > ay && ay > ax then
            ay
        -- -x axis
        else if ay == az && az > ax then
            abs $ nearZero ry rz
        -- IV
        else if ay > az && az > ax then
            az
        -- -y axis
        else if ay > az && az == ax then
            abs $ nearZero rx rz
        -- V
        else if ay > ax && ax > az then
            ax
        -- -z axis
        else if ax == ay && ay > az then
            abs $ nearZero rx ry
        -- VI
        else if ax > ay && ay > az then
            ay
        -- Origin
        else if ax == ay && ay == az then
            ax
        else
            0
    -- -x +y -z
    else if rx <= 0 && ry >= 0 && rz <= 0 then
        -- +x axis
        if ax < az && az == ay then
            - (abs $ nearZero ry rz)
        -- I
        else if ax < az && az < ay then
            -az
        -- +y axis
        else if az == ax && ax < ay then
            - (abs $ nearZero rx rz)
        -- II
        else if az < ax && ax < ay then
            -ax
        -- +z axis
        else if az < ay && ay == ax then
            - (abs $ nearZero rx ry)
        -- III
        else if az < ay && ay < ax then
            -ay
        -- -x axis
        else if ay == az && az < ax then
            - (abs $ nearZero ry rz)
        -- IV
        else if ay < az && az < ax then
            -az
        -- -y axis
        else if ay < az && az == ax then
            - (abs $ nearZero rx rz)
        -- V
        else if ay < ax && ax < az then
            -ax
        -- -z axis
        else if ax == ay && ay < az then
            - (abs $ nearZero rx ry)
        -- VI
        else if ax < ay && ay < az then
            -ay
        -- Origin
        else if ax == ay && ay == az then
            -ax
        else 0
    else 0

nearZero :: (Num a, Ord a) => a -> a -> a
nearZero a b =
    if (abs a) < (abs b) then
        a
    else
        b

File: src/c/loc.c

#include "loc.h"

#include <stdio.h>

void printLoc(char* buf, Loc loc) {
    sprintf(buf, "Loc(%d, %d, %d)", loc.x, loc.y, loc.z);
}

Loc newLoc(int x, int y, int z) {
    Loc loc;
    loc.x = x;
    loc.y = y;
    loc.z = z;
    Loc oloc;
    oloc.x = 0;
    oloc.y = 0;
    oloc.z = 0;
    Loc retLoc = normal(loc, oloc);
    return retLoc;
}

int equalLoc(Loc loc1, Loc loc2) {
    return loc1.x == loc2.x && loc1.y == loc2.y && loc1.z == loc2.z;
}

int distance(Loc l1, Loc l2) {
    Loc l3 = normal(l1, l2);
    return abs(l3.x) + abs(l3.y) + abs(l3.z);
}

Loc normal(Loc l1, Loc l2) {
    int rx = l1.x - l2.x;
    int ry = l1.y - l2.y;
    int rz = l1.z - l2.z;

    int ax = abs(rx);
    int ay = abs(ry);
    int az = abs(rz);
    int delta = getDelta(rx, ry, rz, ax, ay, az);
    Loc retLoc;
    retLoc.x = rx - delta;
    retLoc.y = ry + delta;
    retLoc.z = rz - delta;
    return retLoc;
}

int getDelta(int rx, int ry, int rz, int ax, int ay, int az) {
    // I && II
    // +x +y +z
    if (rx >= 0 && ry >= 0 && rz >= 0) {
        return nearZero(rx, rz);
    // II && III
    // -x +y +z
    } else if (rx <= 0 && ry >= 0 && rz >= 0) {
        int amount = nearZero(rx, ry);
        if (ax > ay) {
            return -amount;
        } else if (ax == ay) {
            return -abs(amount);
        } else {
            return amount;
        }
    // III && IV
    // -x -y +z
    } else if (rx <= 0 && ry <= 0 && rz >= 0) {
        int amount = nearZero(ry, rz);
        if (ay < az) {
            return -amount;
        } else if (ay == az) {
            return abs(amount);
        } else {
            return amount;
        }
    // IV && V
    // -x -y -z
    } else if (rx <= 0 && ry <= 0 && rz <= 0) {
        return nearZero(rx,rz);
    // V && VI
    // +x -y -z
    } else if (rx >= 0 && ry <= 0 && rz <= 0) {
        int amount = nearZero(rx, ry);
        if (ax > ay) {
            return -amount;
        } else if (ax == ay) {
            return abs(amount);
        } else {
            return amount;
        }
    // VI && I
    // +x +y -z
    } else if (rx >= 0 && ry >= 0 && rz <= 0) {
        int amount = nearZero(ry, rz);
        if (ay < az) {
            return -amount;
        } else if (ay == az) {
            return -abs(amount);
        } else {
            return amount;
        }
    // +x -y +z
    } else if (rx >= 0 && ry <= 0 && rz >= 0) {
        // +x axis
        if (ax > az && az == ay) {
            return abs(nearZero(ry,rz));
        // I
        } else if (ax > az && az > ay) {
            return az;
        // +y axis
        } else if (az == ax && ax > ay) {
            return abs(nearZero(rx,rz));
        // II
        } else if (az > ax && ax > ay) {
            return ax;
        // +z axis
        } else if (az > ay && ay == ax) {
            return abs(nearZero(rx,ry));
        // III
        } else if (az > ay && ay > ax) {
            return ay;
        // -x axis
        } else if (ay == az && az > ax) {
            return abs(nearZero(ry,rz));
        // IV
        } else if (ay > az && az > ax) {
            return az;
        // -y axis
        } else if (ay > az && az == ax) {
            return abs(nearZero(rx,rz));
        // V
        } else if (ay > ax && ax > az) {
            return ax;
        // -z axis
        } else if (ax == ay && ay > az) {
            return abs(nearZero(rx,ry));
        // VI
        } else if (ax > ay && ay > az) {
            return ay;
        // Origin
        } else if (ax == ay && ay == az) {
            return ax;
        } else {
            return 0;
        }
    // -x +y -z
    } else if (rx <= 0 && ry >= 0 && rz <= 0) {
        // +x axis
        if (ax < az && az == ay) {
            return -abs(nearZero(ry,rz));
        // I
        } else if (ax < az && az < ay) {
            return -az;
        // +y axis
        } else if (az == ax && ax < ay) {
            return -abs(nearZero(rx,rz));
        // II
        } else if (az < ax && ax < ay) {
            return -ax;
        // +z axis
        } else if (az < ay && ay == ax) {
            return -abs(nearZero(rx,ry));
        // III
        } else if (az < ay && ay < ax) {
            return -ay;
        // -x axis
        } else if (ay == az && az < ax) {
            return -abs(nearZero(ry,rz));
        // IV
        } else if (ay < az && az < ax) {
            return -az;
        // -y axis
        } else if (ay < az && az == ax) {
            return -abs(nearZero(rx,rz));
        // V
        } else if (ay < ax && ax < az) {
            return -ax;
        // -z axis
        } else if (ax == ay && ay < az) {
            return -abs(nearZero(rx,ry));
        // VI
        } else if (ax < ay && ay < az) {
            return -ay;
        // Origin
        } else if (ax == ay && ay == az) {
            return -ax;
        } else {
            return 0;
        }
    } else {
        return 0;
    }
}

int nearZero(int a, int b) {
    if (abs(a) < abs(b)) {
        return a;
    } else {
        return b;
    }
}
